wgpu-core/src/instance.rs

```diff
@@ -555,64 +453,39 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
-            if selected < adapters_vk.len() {
-                let adapter = Adapter::new(adapters_vk.swap_remove(selected), unsafe_extensions);
-                log::info!("Adapter Vulkan {:?}", adapter.raw.info);
-                let id = backend::Vulkan::hub(self).adapters.register_identity(
-                    id_vulkan.unwrap(),
-                    adapter,
-                    &mut token,
-                );
-                return Some(id);
-            }
-            selected -= adapters_vk.len();
-        }
-        #[cfg(any(target_os = "ios", target_os = "macos"))]
-        {
-            if selected < adapters_mtl.len() {
-                let adapter = Adapter::new(adapters_mtl.swap_remove(selected), unsafe_extensions);
-                log::info!("Adapter Metal {:?}", adapter.raw.info);
-                let id = backend::Metal::hub(self).adapters.register_identity(
-                    id_metal.unwrap(),
-                    adapter,
-                    &mut token,
-                );
-                return Some(id);
-            }
-            selected -= adapters_mtl.len();
-        }
-        #[cfg(windows)]
-        {
-            if selected < adapters_dx12.len() {
-                let adapter = Adapter::new(adapters_dx12.swap_remove(selected), unsafe_extensions);
-                log::info!("Adapter Dx12 {:?}", adapter.raw.info);
-                let id = backend::Dx12::hub(self).adapters.register_identity(
-                    id_dx12.unwrap(),
-                    adapter,
-                    &mut token,
-                );
-                return Some(id);
-            }
-            selected -= adapters_dx12.len();
-            if selected < adapters_dx11.len() {
-                let adapter = Adapter::new(adapters_dx11.swap_remove(selected), unsafe_extensions);
-                log::info!("Adapter Dx11 {:?}", adapter.raw.info);
-                let id = backend::Dx11::hub(self).adapters.register_identity(
-                    id_dx11.unwrap(),
-                    adapter,
-                    &mut token,
-                );
-                return Some(id);
-            }
-            selected -= adapters_dx11.len();
+
+        backends_map! {
+            let map = |(info_adapter, id_backend, mut adapters_backend, hub)| {
+                if selected < adapters_backend.len() {
+                    let adapter = Adapter::new(adapters_backend.swap_remove(selected), unsafe_extensions);
+                    log::info!("Adapter {} {:?}", info_adapter, adapter.raw.info);
+                    let id = hub().adapters.register_identity(
+                        id_backend.take().unwrap(),
+                        adapter,
+                        &mut token,
+                    );
+                    return Some(id);
+                }
+                selected -= adapters_backend.len();
+            };
+
+            #[vulkan]
+            map(("Vulkan", &mut id_vulkan, adapters_vk, || backend::Vulkan::hub(self))),
```

Comment:
can we pass it as `backend::Vulkan::hub`? again, it's a static function with known argument, which `map()` can call on `self`

Reply:
Sure, I'll change it.